merlin/dispatch/lazy.py

```diff
@@ -0,0 +1,131 @@
+
+
+class LazyDispatcher:
+    """
+    A wrapper around `functools.singledispatch` that allows lazy registration
+    if the top-level module containing the types to be registered becomes
+    available later.
+    """
+
+    def __init__(self, func):
+        self.dispatcher = singledispatch(func)
+        self.func = func
+        self._lazy = {}
+
+    def register(self, cls, func=None):
+        """
+        Registers a new implementation for the given *cls* on a *generic_func*.
+
+        generic_func.register(cls, func) -> func
+
+        Parameters
+        ----------
+        func : function, optional
+            The registration function to be lazily executed if the class
+            becomes available, by default None
+
+        Returns
+        -------
+        function
+            Version of func wrapped by `functools.singledispatch`
+        """
+        return self.dispatcher.register(cls, func=func)
+
+    def register_lazy(self, module_name: str, func=None):
+        """
+        Registers a new lazy registration function for the given *module* on a *generic_func*.
+
+        To use lazy registration, the default implementation of the generic function *must*
+        raise NotImplementedError.
+
+        Parameters
+        ----------
+        module_name : str
+            Name of the top level module that will be imported in the registration func.
+        func : function, optional
+            The registration function to be lazily executed if the top level module
+            becomes available, by default None
+
+        Returns
+        -------
+        function
+            Wrapped version of func
+        """
+
+        def wrapper(func):
+            self._lazy[module_name] = func
+            return func
+
+        return wrapper(func) if func is not None else wrapper
+
+    def __call__(self, *args_, **kwargs_):
+        try:
+            dispatch_type = self._dispatch_type(args_[0])
+            return self._singledispatch(dispatch_type, *args_, **kwargs_)
+        except NotImplementedError:
+            try:
+                module_name = dispatch_type.__module__.partition(".")[0]
+                self._lazy[module_name]()
+                self._lazy.pop(module_name, None)
+            except (AttributeError, KeyError):
+                self._raise_not_impl(self.func, args_[0])
+            else:
+                return self._singledispatch(dispatch_type, *args_, **kwargs_)
+
+    def _singledispatch(self, obj_type, *args, **kwargs):
+        return self.dispatcher.dispatch(obj_type)(*args, **kwargs)
+
+    def _dispatch_type(self, arg):
+        return arg if isclass(arg) else type(arg)
+
+    def _raise_not_impl(self, func, arg):
+        funcname = getattr(func, "__name__", "lazysingledispatch function")
+        raise NotImplementedError(
+            f"{funcname} doesn't have a registered implementation "
+            f"for type of {arg} ({type(arg)})"
+        )
+
+
+def lazysingledispatch(func):
```

Comment:
The name of this function feels a bit busy to me. It probably doesn't need to include the "lazy" detail (and maybe not even "single"). I don't have very strong feelings, but something as simple as `dispatchable` feels a bit nicer.

Reply:
>That said, I don't feel very strongly about this, because I can see the logic of one and the aesthetic appeal of the other.

I'm probably in the same boat. Though, I guess I'd probably be a slightly happier with `lazy_singledispatch` than `lazysingledispatch`. I think the root of the problem is that I'm just annoyed that functools` choice of `singledispatch` over `single_dispatch`. Something about three words concatenated together just stresses my brain a bit too much.

Reply:
I think that's reasonable and agree that an underscore may improve readability here 👍🏻